## Практическое задание

Попробуйте обучить, нейронную сеть на Keras(рассмотренную на уроке) на датасете MNIST с другими параметрами. Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?

Поработайте с документацией Keras. Попробуйте найти полезные команды Keras неразобранные на уроке.

## Базовая модель 

In [8]:
# The full neural network code!
###############################
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')

# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

Epoch 1/5
1875/1875 [==============================] - 1s 581us/step - loss: 0.3585 - accuracy: 0.8934
Epoch 2/5
1875/1875 [==============================] - 1s 573us/step - loss: 0.1890 - accuracy: 0.9431
Epoch 3/5
1875/1875 [==============================] - 1s 560us/step - loss: 0.1476 - accuracy: 0.9543
Epoch 4/5
1875/1875 [==============================] - 1s 571us/step - loss: 0.1236 - accuracy: 0.9625
Epoch 5/5
313/313 [==============================] - 0s 477us/step - loss: 0.1393 - accuracy: 0.9568
[7 2 1 0 4]
[7 2 1 0 4]


## Улучшенная модель 

1. Добавим один слой 
2. Увеличим количество эпох

In [11]:
# The full neural network code!
###############################
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

# Compile the model.
model.compile(
  optimizer='adam',
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=10,
  batch_size=32,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model_better.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')

# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

Epoch 1/10
1875/1875 [==============================] - 1s 718us/step - loss: 0.3613 - accuracy: 0.8892
Epoch 2/10
1875/1875 [==============================] - 1s 613us/step - loss: 0.1789 - accuracy: 0.9451
Epoch 3/10
1875/1875 [==============================] - 1s 609us/step - loss: 0.1369 - accuracy: 0.9573
Epoch 4/10
1875/1875 [==============================] - 1s 643us/step - loss: 0.1154 - accuracy: 0.9641
Epoch 5/10
1875/1875 [==============================] - 1s 655us/step - loss: 0.1010 - accuracy: 0.9683
Epoch 6/10
1875/1875 [==============================] - 1s 606us/step - loss: 0.0914 - accuracy: 0.9717
Epoch 7/10
1875/1875 [==============================] - 1s 601us/step - loss: 0.0824 - accuracy: 0.9742
Epoch 8/10
1875/1875 [==============================] - 1s 633us/step - loss: 0.0742 - accuracy: 0.9763
Epoch 9/10
1875/1875 [==============================] - 1s 606us/step - loss: 0.0690 - accuracy: 0.9774
Epoch 10/10
313/313 [==============================] - 0s 509us/

## Резюме:
    
Увеличили точность с accuracy: 0.9568 до  0.9713 и сократили loss c 0.1393 до 0.0969 за счет добавления дополнительного слоя и увеличения количества эпох. Посмотрел cheat cheet по Keras, каких-то особых находок не обнаружил